### DISCUSSION TITLE
(WIP) RFC-023: On-demand implement stats

### DISCUSSION THREAD

**Discussion by ahabhgk** [web-infra-dev/rspack#837] [RFC]
- Feature Name: Stats
- Start Date: 2022-10-12
- Rspack Issue: [speedy-js/rspack#0000](https://github.com/speedy-js/rspack/issues/0000) Issue name should be `Tracking issue of <Fill in the RFC title above>`

# Summary
[summary]: #summary

The stats option lets you precisely control what bundle information gets displayed.

# Motivation
[motivation]: #motivation

Let users can use tools like `webpack-bundle-analyzer` to analyze their bundle information, and optimize their bundles.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Stats is about getting bundle information from compilation.

Since there are too lots of fields in webpack stats json and have strong relation with its compilation data and methods, port all of these fields to rspack stats is hard and needs rspack compilation implementation highly aligned with webpack compilation.

So fields in rspack stats will be implemented **on-demand**, and listing which fields needs implemented with demand in the comments is encouraged.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

```rust
#[derive(Debug, Clone)]
pub struct Stats<'compilation> {
  compilation: &'compilation Compilation,
}

#[derive(Debug, Clone)]
pub struct StatsDescription {
  pub assets: Vec<StatsAsset>,
  pub modules: Vec<StatsModule>,
  pub chunks: Vec<StatsChunk>,
  pub errors: Vec<StatsError>,
  pub errors_count: u32,
  // ...
}

impl Stats {
  // align with `Stats.toJson` in webpack
  pub fn to_description(&self) -> StatsDescription { /* ... */ }
}

// align with `Stats.toString` in webpack
impl Display for StatsDescription { /* ... */ }
```

use options to control which field is included in stats json will be implemented in js side.

# Prior art
[prior-art]: #prior-art

Stats in webpack has three concept:

1. options: specify which information you want to see in the stats output.
2. factory: customizable stats json creator with hooks.
3. printer: customizable stats string creator with hooks.

```
Stats.toJson: options --> statsFactory --> stats json
Stats.toString: options --> statsFactory --> stats json --> statsPrinter --> stats string
```

The main idea about stats in webpack is still getting bundle information from compilation, but it can be customized by its hooks ([which is added in webpack5](https://github.com/webpack/webpack/blob/v4.46.0/lib/Stats.js#L104) and not used widely).

**Comment by ahabhgk** [comment]
# Webpack Bundle Analyzer

Only needs few fields of stats, but have some limit for now.

fields:

- `assets.type`
- `assets.name`
- `assets.size`
- `assets.chunks`
- `modules.type`
- `modules.moduleType`
- `modules.identifier`
- `modules.name`
- `modules.id`
- `modules.chunks`
- `modules.size`

limits:

- require the output format of rspack align with webpack, cause it uses `acorn` to parse the output asset to find out what modules used in asset.
- the runtime should not as a file in all chunks, it should be set as `Entrypoint.runtimeChunk` or a independent chunk when [`entry.runtime`](https://webpack.js.org/concepts/entry-points#:~:text=runtime%3A%20The%20name%20of%20the%20runtime%20chunk.%20When%20set%2C%20a%20new%20runtime%20chunk%20will%20be%20created.%20It%20can%20be%20set%20to%20false%20to%20avoid%20a%20new%20runtime%20chunk%20since%20webpack%205.43.0.) has value or [optimization.runtimeChunk](https://webpack.js.org/configuration/optimization/#optimizationruntimechunk) is used.
- `modules.name` should be a relative path to context (`./src/index.js`) rather than an absolute path.

**Comment by hardfist** [comment]
It would be better if you could summarize the bundle analyzer implementation

**Comment by ahabhgk** [comment]
# [BundleStats](https://github.com/relative-ci/bundle-stats)

fields: [bundle-stats/packages/plugin-webpack-filter/src/index.ts](https://github.com/relative-ci/bundle-stats/blob/94fc08fac0ecc6c9288bee7f8b15045e6bafaabe/packages/plugin-webpack-filter/src/index.ts)
